crates/ide_assists/src/handlers/generate_default_from_new.rs

```diff
@@ -0,0 +1,364 @@
+use crate::{
+    assist_context::{AssistContext, Assists},
+    AssistId,
+};
+use hir::TypeRef;
+use syntax::{
+    ast::{self, Impl, NameOwner},
+    AstNode,
+};
+use test_utils::mark;
+
+// Assist: generate_default_from_new
+//
+// Generates default implementation from new method.
+//
+// ```
+// struct Example { _inner: () }
+//
+// impl Example {
+//     pub fn n$0ew() -> Self {
+//         Self { _inner: () }
+//     }
+// }
+// ```
+// ->
+// ```
+// struct Example { _inner: () }
+//
+// impl Example {
+//     pub fn new() -> Self {
+//         Self { _inner: () }
+//     }
+// }
+//
+// impl Default for Example {
+//     fn default() -> Self {
+//         Self::new()
+//     }
+// }
+// ```
+pub(crate) fn generate_default_from_new(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let fn_node = ctx.find_node_at_offset::<ast::Fn>()?;
+    let fn_name = fn_node.name()?;
+
+    if fn_name.text() != "new" {
+        mark::hit!(other_function_than_new);
+        return None;
+    }
+
+    if fn_node.param_list()?.params().next().is_some() {
+        mark::hit!(new_function_with_parameters);
+        return None;
+    }
+
+    let impl_ = fn_node.syntax().ancestors().into_iter().find_map(ast::Impl::cast)?;
+    if is_default_implemented(ctx, &impl_).is_some() {
+        return None;
+    }
+
+    let insert_location = impl_.syntax().text_range();
+
+    acc.add(
+        AssistId("generate_default_from_new", crate::AssistKind::Generate),
+        "Generate a Default impl from a new fn",
+        insert_location,
+        move |builder| {
+            let code = default_fn_node_for_new(impl_);
+            builder.insert(insert_location.end(), code);
+        },
+    )
+}
+
+fn default_fn_node_for_new(impl_: Impl) -> String {
+    format!(
+        "
+
+impl Default for {} {{
+    fn default() -> Self {{
+        Self::new()
+    }}
+}}",
+        impl_.self_ty().unwrap().syntax().text()
+    )
+}
+
+fn is_default_implemented(ctx: &AssistContext, impl_: &Impl) -> Option<bool> {
+    let db = ctx.sema.db;
+    let module = impl_.syntax().parent()?;
+    let sema_scope = ctx.sema.scope(&module);
+    let impls = sema_scope.module()?.impl_defs(db);
+    let mut name = None;
+    for i in impls {
+        if let Some(TypeRef::Path(p)) = i.target_trait(db) {
+            name = p.segments().iter().map(|s| s.name.to_string()).find(|n| n == "Default");
+        }
+    }
+
+    name.map(|n| !n.is_empty())
+}
```

Comment:
That's not qutie the right approach, we want to make use of the HIR for this and ask the type directly whether it implements Default, you can do that roughly like so:
```rust
let impl_: ast::Impl = ...;
let impl_ = ctx.sema.to_def(&impl_)?;
let ty = impl_.target_ty(ctx.sema.db);
let krate = ty.module(sema.db).krate();
let default_trait = FamousDefs(&ctx.sema, Some(krate)).core_default_Default()?;
let implements_default = ty.impls_trait(sema.db, default_trait, &[]);
```

Reply:
Hi @Veykril ,
I have implemented below code. but for Default imlementation it is not giving true.

fn is_default_implemented(ctx: &AssistContext, impl_: &Impl) -> Option<bool> {
    let db = ctx.sema.db;
    let impl_def = ctx.sema.to_def(impl_)?;
    let ty = impl_def.target_ty(db);
    let krate = impl_def.module(db).krate();
    let default_trait = FamousDefs(&ctx.sema, Some(krate)).core_default_Default()?;
    let implements_default = ty.impls_trait(db, default_trait, &[]);
    Some(implements_default)
}

Reply:
I think you misspelled one of the imp blocks. In two of your tests you accidentally wrote `Exmaple` instead of `Example`